R/mediate.R

```diff
@@ -0,0 +1,386 @@
+      df[[covariate]] <- colData(se)[[covariate]]
+      
+      # Add covariate to formula of mediation model
+      relation_m <- paste(relation_m, "+", covariate)
+      # Add covariate to formula of outcome model
+      relation_dv <- paste(relation_dv, "+", covariate)
+    }
+  }
+
+  # Fit mediation model
+  fit_m <- do.call(lm, list(formula = formula(relation_m),
+                            data = df))
+  # Fit outcome model
+  fit_dv <- do.call(glm, list(formula = formula(relation_dv),
+                              family = family,
+                              data = df))
+  # Run mediation analysis
+  med_out <- mediate(fit_m, fit_dv,
+                     treat = "Treatment",
+                     mediator = "Mediator",
+                     covariates = covariates,
+                     ...)
+  
+  return(med_out)
+}
+
+
+# Update list of results
+update.results <- function(results, med_out,
+                           treatment, mediator, outcome) {
+  
+  # Update model variables
+  results[["Treatment"]] <- c(results[["Treatment"]], treatment)
+  results[["Mediator"]] <- c(results[["Mediator"]], mediator)
+  results[["Outcome"]] <- c(results[["Outcome"]], outcome)
+  
+  # Update stats of ACME (average causal mediation effect)
+  results[["ACME_estimate"]] <- c(results[["ACME_estimate"]], med_out$d.avg)
+  results[["ACME_pval"]] <- c(results[["ACME_pval"]], med_out$d.avg.p)
+  results[["ACME_ci"]] <- c(results[["ACME_ci"]], med_out$d.avg.ci)
+  
+  # Update stats of ADE (average direct effect)
+  results[["ADE_estimate"]] <- c(results[["ADE_estimate"]], med_out$z.avg)
+  results[["ADE_pval"]] <- c(results[["ADE_pval"]], med_out$z.avg.p)
+  results[["ADE_ci"]] <- c(results[["ADE_ci"]], med_out$z.avg.ci)
+  
+  # Add current model to metadata
+  results[["Model"]][[length(results[["Model"]]) + 1]] <- med_out
+  
+  return(results)
+}
+
+
+# Combine results into output dataframe
+make.output <- function(results, p.adj.method, add.metadata) {
+  
+  # Create dataframe with model variables, effect sizes and p-values
+  med_df <- do.call(data.frame, results[1:(length(results) - 3)])
+  
+  # Compute adjusted p-values and add them to dataframe
+  med_df[["ACME_adjpval"]] <- p.adjust(med_df[["ACME_pval"]], method = p.adj.method)
+  med_df[["ADE_adjpval"]] <- p.adjust(med_df[["ADE_pval"]], method = p.adj.method)
+  
+  # Split CI lists into lower and upper limits and add them to dataframe
+  med_df[["ACME_CI_lower"]] <- results[["ACME_ci"]][names(results[["ACME_ci"]]) == "2.5%"]
+  med_df[["ACME_CI_upper"]] <- results[["ACME_ci"]][names(results[["ACME_ci"]]) == "97.5%"]
+  med_df[["ADE_CI_lower"]] <- results[["ADE_ci"]][names(results[["ADE_ci"]]) == "2.5%"]
+  med_df[["ADE_CI_upper"]] <- results[["ADE_ci"]][names(results[["ADE_ci"]]) == "97.5%"]
+  
+  if (add.metadata) {
+    # If desired, the models for every mediator are saved into the metadata attribute
+    attr(med_df, "metadata") <- results[["Model"]]
```

Comment:
It looks to me that .add_values_to_metadata takes a (Tree)SE object as input. However, here the models are added to the output dataframe, not the original se object.

Reply:
Do we have a reason not to add them to (Tree)SE? Is it necessary to return them to some dataframe instead of TreeSE metadataslot - or is this some intermediate step?

Reply:
1.

My mistake. I looked the code now better. The returned value is now df.


2.

We could return data.frame in `metadata(tse)`

cons
-  The resulting df is not linked anywhere i.e., the columns are not linked with `colnames(tse)` or `rownames(tse)`
- User only needs the df and fetching it from `metadata(tse)` might feel like additional step

pros
+ We have other functions that add the results to `metadata(tse)` (like `cluster` adds detailed clustering info)
+ Adding it to `metadata(tse)` is "correct" way to use the data container
+ Calculating might take long time and by storing it to `metadata(tse)` keeps it safe there

As this is new function, we can decide the output type freely. There could be `as.df` option, but I think results would be better to store in `metadata(tse)` by default

Reply:
What if we have a function called `getMediation`, which returns the dataframe object, and a function called `runMediation`, which returns the TreeSE with df as metadata?